### PR TITLE
Advanced postcss plugins cli

### PR DESCRIPTION
**Pull request by cef62**
This PR address #12 adding support for loading post-css plugins configuration via an external module.
The CLI expect a path to the module, the module must export a function returning an array of post-css modules.
I've also added test for the CLI, only test for post-css plugins are missing,

Some notes:
- CLI test are really slow, because they really run the whole compiler
- I've tried to mock the environment to avoid the complete execution of the module, we should just test arguments parsing but I've not been successful. I0've tried using babel-rewire. @MicheleBertoli any suggestion/solution it will be really great!

I've also added unit test code coverage using [nyc](https://github.com/istanbuljs/nyc), it works but I'm not able to make it works with the emojis reporter. Using the reporter an error is thrown at the end of the tests. Is not blocking but it's annoying.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,6 +6,8 @@
   "description": "An automation task to compile css-modules and remove it from the components",
   "scripts": {
     "test": "NODE_ENV=test mocha --compilers js:babel-register --reporter emoji-reporter ./test/**/*.spec.js",
+    "coverage": "NODE_ENV=test nyc --cache --require babel-core/register mocha ./test/**/*.spec.js --timeout=8000",
```
Comment (MicheleBertoli) [inline package.json]: I tried to run this command as `NODE_ENV=test nyc npm test` and I didn't get any errors, can you please give me more info about the emoji issue?

**Comment by MicheleBertoli** [comment]
Well done!

I would really split the `cli-parser` (and its tests) into separate modules.
If you don't like the idea, at least we should extract `getPlugins` (and test it).

File: package.json (hunk 1 above)
Comment (cef62) [inline package.json]: You're right, I think the problem is in how I'm invoking mocha trough `nyc`. I've updated the scripts [here](https://github.com/cef62/css-modules-compiler/pull/23/commits/6261b166b4ff657d963936e6635099871e5950e0) and it works perfectly now!

**Comment by cef62** [comment]
I don't mind splitting `cli-parser`, I like the idea! I'll do it as soon as possible ;)

File: src/cli-parser/index.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const handler = require('./handler')
+const builder = require('./builder')
+const command = 'compile'
+const describe = 'Compile css-modukes to a single css file and update css import in Javascript'
+
+
```
Comment (MicheleBertoli) [inline src/cli-parser/index.js]: Double new line

File: src/cli-parser/index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+exports.builder = builder
+exports.handler = handler
+exports.command = command
+exports.describe = describe
+// module.exports = { builder, command, describe, handler }
```
Comment (MicheleBertoli) [inline src/cli-parser/index.js]: Comment

File: test/cli/index.spec.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable no-underscore-dangle */
+import assert from 'assert'
+import { builder, command, describe as describeCommand, handler } from '../../src/cli-parser'
+
+describe('CLI parser - .command', () => {
```
Comment (MicheleBertoli) [inline test/cli/index.spec.js]: What about nested describes?

File: test/cli/handler.spec.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+const getPluginsEmpty = createSpy(() => {})
+
+const error = createSpy()
+const debug = createSpy()
+
+handler.__Rewire__('compileCss', compiler)
+handler.__Rewire__('error', error)
+handler.__Rewire__('debug', debug)
+
+describe('CLI - .handler()', () => {
```
Comment (MicheleBertoli) [inline test/cli/handler.spec.js]: What about just "handler"?

File: test/cli/builder.spec.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/* eslint-disable no-underscore-dangle */
+import assert from 'assert'
+import { yargsMock } from './helpers'
+import builder from '../../src/cli-parser/builder'
+
+const yargs = yargsMock
+
+describe('CLI parser - .builder()', () => {
```
Comment (MicheleBertoli) [inline test/cli/builder.spec.js]: What about just "builder"?

**Comment by MicheleBertoli** [comment]
AWESOME!

Just a few small inline comments and a couple of questions:
- wouldn't it be better to rename the `cli` folder (in tests) to `cli-parser` to be consistent with the structure of `src`?
- the `get-plugins` tests are gone but since you are mocking the module in the `handler` specs it would be good to add a couple of tests for it, what do you think?

Obviously, feel free to merge the PR. We make this changes later.

**Comment by cef62** [comment]
I'm not sure about removing 'CLI parser -' from `builder` and `handler` specs. I know we use emojis :P but I kind of like prefixing the context. Not a big deal anyway :)

> the get-plugins tests are gone but since you are mocking the module in the handler specs it would be good to add a couple of tests for it, what do you think

Yes `get-plugins` must be tested, I forget to add its test while refactoring ;)

**Comment by MicheleBertoli** [comment]
Got it @cef62, in that case you should use nested describes:

``` javascript
describe('CLI parse', () => {
  describe('builder', () => {
    it('works', () => {
      ...
    })
  })
})
```

I would do that in `test/cli/index.spec.js` too because `it('.command should` sounds weird.

**Comment by cef62** [comment]
Done, now should be decent.
If it's ok for you I'll merge it tomorrow.

**Comment by MicheleBertoli** [comment]
Perfect!

giphy 1

**Pull request merged**